src/framework/components/anim/controller.js

```diff
@@ -4,17 +4,295 @@ import {
     ANIM_GREATER_THAN, ANIM_LESS_THAN, ANIM_GREATER_THAN_EQUAL_TO, ANIM_LESS_THAN_EQUAL_TO, ANIM_EQUAL_TO, ANIM_NOT_EQUAL_TO,
     ANIM_INTERRUPTION_NONE, ANIM_INTERRUPTION_PREV, ANIM_INTERRUPTION_NEXT, ANIM_INTERRUPTION_PREV_NEXT, ANIM_INTERRUPTION_NEXT_PREV,
     ANIM_PARAMETER_TRIGGER,
-    ANIM_STATE_START, ANIM_STATE_END, ANIM_STATE_ANY
+    ANIM_STATE_START, ANIM_STATE_END, ANIM_STATE_ANY,
+    ANIM_BLEND_1D, ANIM_BLEND_2D_DIRECTIONAL, ANIM_BLEND_2D_CARTESIAN, ANIM_BLEND_DIRECT
 } from './constants.js';
 
-function AnimState(controller, name, speed, loop) {
+function AnimNode(parent, name, point, speed) {
+    this._parent = parent;
+    this._name = name;
+    this._point = Array.isArray(point) ? new pc.Vec2(point) : point;
+    this._speed = speed || 1.0;
+    this._weight = 1.0;
+    this._animTrack = null;
+}
+
+Object.defineProperties(AnimNode.prototype, {
+    parent: {
+        get: function () {
+            return this._parent;
+        }
+    },
+    name: {
+        get: function () {
+            return this._name;
+        }
+    },
+    path: {
+        get: function () {
+            return this._parent ? this._parent.path + '.' + this._name : this._name;
+        }
+    },
+    point: {
+        get: function () {
+            return this._point;
+        }
+    },
+    weight: {
+        get: function () {
+            return this._parent ? this._parent.weight * this._weight : this._weight;
+        },
+        set: function (value) {
+            this._weight = value;
+        }
+    },
+    speed: {
+        get: function () {
+            return this._speed;
+        }
+    },
+    animTrack: {
+        get: function () {
+            return this._animTrack;
+        },
+        set: function (value) {
+            this._animTrack = value;
+        }
+    }
+});
+
+function BlendTree(parent, name, point, type, parameters, children, findParameter) {
```

Comment:
Again, just a few sentences explaining the purpose of this class in the system would be great!

Reply:
👍